### PR TITLE
Add V2 CapacityGroup REST API to TitusFederation

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-server-federation/src/main/java/com/netflix/titus/federation/endpoint/rest/FederationV2CapacityGroupResource.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import com.netflix.titus.api.endpoint.v2.rest.representation.ApplicationSlaRepresentation;
+import com.netflix.titus.common.util.tuple.Either;
+import com.netflix.titus.federation.endpoint.EndpointConfiguration;
+import com.netflix.titus.federation.service.CellWebClientConnector;
+import com.netflix.titus.federation.service.CellWebClientConnectorUtil;
+import io.swagger.annotations.Api;
+import org.springframework.core.ParameterizedTypeReference;
+
+/**
+ * Supports subset of operations provided by {@link com.netflix.titus.grpc.protogen.SchedulerServiceGrpc}, applicable
```
Comment (fabiokung) [inline titus-server-federation/src/main/java/com/netflix/titus/federation/endpoint/rest/FederationV2CapacityGroupResource.java]: `SchedulerServiceGrpc` copy and paste?

File: titus-server-federation/src/main/java/com/netflix/titus/federation/endpoint/rest/FederationV2CapacityGroupResource.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    private static final ParameterizedTypeReference<List<ApplicationSlaRepresentation>> APPLICATION_SLA_LIST_TP =
+            new ParameterizedTypeReference<List<ApplicationSlaRepresentation>>() {
+            };
+
+    private final EndpointConfiguration configuration;
+    private final CellWebClientConnector cellWebClientsSupplier;
+
+    @Inject
+    public FederationV2CapacityGroupResource(EndpointConfiguration configuration,
+                                             CellWebClientConnector cellWebClientsSupplier) {
```
Comment (fabiokung) [inline titus-server-federation/src/main/java/com/netflix/titus/federation/endpoint/rest/FederationV2CapacityGroupResource.java]: minor nit: rename variable and field, it looks like this was a `Supplier<...>` before?

File: titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnectorUtil.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import org.springframework.web.reactive.function.client.WebClientResponseException;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class CellWebClientConnectorUtil {
+
+    /**
+     * Run GET operation on all cells and merge the result.
+     */
+    public static <T> Either<List<T>, WebApplicationException> doGetAndMerge(CellWebClientConnector cellWebClientsSupplier,
```
Comment (fabiokung) [inline titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnectorUtil.java]: nit: remove supplier suffix

File: titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnectorUtil.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                    .collectList()
+                    .<Either<List<List<T>>, Throwable>>map(Either::ofValue)
+                    .onErrorResume(e -> Mono.just(Either.ofError(e)))
+                    .block(Duration.ofMillis(timeoutMs));
+        } catch (Exception e) {
+            return Either.ofError(new WebApplicationException(e, Response.Status.INTERNAL_SERVER_ERROR));
+        }
+
+        if (partials == null) {
+            return Either.ofError(new WebApplicationException(Response.Status.INTERNAL_SERVER_ERROR));
```
Comment (fabiokung) [inline titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnectorUtil.java]: minor: add an error message (e.g. `No results from any cell`) to aid debugging in the future

File: titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnector.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import com.netflix.titus.api.federation.model.Cell;
+import org.springframework.web.reactive.function.client.WebClient;
+
+public interface CellWebClientConnector {
+
+    interface WebClientFactory extends Function<Cell, WebClient> {
```
Comment (fabiokung) [inline titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnector.java]: I think we need a default implementation of this for the OSS packaging, no? (not the Embedded one used in tests, but one for `Main` classes of Federation). Without one, I'd expect the `docker-compose` test to fail

Comment (fabiokung) [reply titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnector.java]: ah, maybe we are not installing a federation instance in the `docker-compose` test

File: titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnectorUtil.java (hunk 4 above)
Comment (tbak) [reply titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnectorUtil.java]: I will add some log lines.

File: titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnector.java (hunk 5 above)
Comment (tbak) [reply titus-server-federation/src/main/java/com/netflix/titus/federation/service/CellWebClientConnector.java]: I have added a simple implementation.

**Pull request merged**